albatross/core/traits.h

```diff
@@ -28,36 +28,82 @@ namespace albatross {
  */
 template <class T> struct delay_static_assert : std::false_type {};
 
+/*
+ * In CovarianceFunction we frequently inspect for definitions of
+ * call_impl_ which MUST be defined for const references to objects
+ * (so that repeated covariance matrix evaluations return the same thing
+ *  and so the computations are not repeatedly copying.)
+ * This type conversion utility will turn a type `T` into `const T&`
+ */
+template <class T> struct call_impl_arg_type {
+  typedef
+      typename std::add_lvalue_reference<typename std::add_const<T>::type>::type
+          type;
+};
+
 /*
  * This determines whether or not a class has a method defined for,
  *   `operator() (X x, Y y, Z z, ...)`
  * The result of the inspection gets stored in the member `value`.
  */
 template <typename T, typename... Args> class has_call_operator {
+
   template <typename C,
-            typename = decltype(std::declval<C>()(std::declval<Args>()...))>
-  static std::true_type test(int);
-  template <typename C> static std::false_type test(...);
+            typename = decltype(std::declval<C>()(
+                std::declval<typename call_impl_arg_type<Args>::type>()...))>
+  static std::true_type test(C *);
+  template <typename> static std::false_type test(...);
 
 public:
   static constexpr bool value = decltype(test<T>(0))::value;
 };
 
 /*
  * This determines whether or not a class has a method defined for,
- *   `call_impl_(X x, Y y, Z z, ...)`
+ *   `double call_impl_(const X &x, const Y &y, const Z &z, ...)`
  * The result of the inspection gets stored in the member `value`.
  */
-template <typename T, typename... Args> class has_defined_call_impl {
+template <typename T, typename... Args> class has_valid_call_impl {
+
+  template <typename C>
+  static auto test(C *) -> typename std::is_same<
```

Comment:
Any particular reason you changed how you are defining the `test()` function trait thing? Everywhere else you have your SFINAE switch in the template and here it's a part of the return type declaration.

Reply:
No great reason, I'll swtich back for consistency.